### PR TITLE
Set `DAEMON_VERSION` from git metadata

### PR DESCRIPTION
**Pull request by git-developer**
This is a proposal for https://github.com/C0rn3j/sc-controller/issues/14 that reads the `DAEMON_VERSION` from git metadata and replaces the value of the constant at build time.

With this approach, the [constant in constants.py](https://github.com/C0rn3j/sc-controller/blob/037a5d65d11aa6358ea5cc4c788bb0c64682fbf4/scc/constants.py#L29) is not versioned in git. It could be changed to something like `dev`, `latest`, `local` or the like.

When the `GIT_DESCRIPTION` build-arg is set, the daemon version is determined automatically. Alternatively, the build-arg `DAEMON_VERSION` may be set explicitely (merely for debugging and local experiments).

#### Examples
| Git Ref | Git Description | `DAEMON_VERSION` |
|---|---|---|
| v0.4.9.2 | v0.4.9.2 | 0.4.9.2 |
| v0.4.9.2^ | v0.4.9.1.1-8-g9f7dc8b | 0.4.9.1.1.8.dev167238795 |
| images | v0.4.8.16-2-g414fc45 | 0.4.8.16.2.dev68484165 |

#### References
* [git describe](https://git-scm.com/docs/git-describe) explains the git metadata format
* [version-specifiers]( https://packaging.python.org/en/latest/specifications/version-specifiers/#version-specifiers) explains the python version format (PEP 440)
* [Example build](https://github.com/git-developer/sc-controller/actions/runs/10983237767#artifacts) with artifacts

### CODE REVIEW

**Comment by git-developer** [comment]
Idea: maybe the version constant can be externalized to an .env file so that `constants.py` is not touched.

**Comment by C0rn3j** [comment]
So basically, it will only remain broken for building from git sources directly(where we could somehow set version from git tag instead, ideally)?

The releases will have AppImages with correct versions, and sources with the correct versions too as per their release tags?

**Comment by git-developer** [comment]
> So basically, it will only remain broken for building from git sources directly(where we could somehow set version from git tag instead, ideally)?

When the build is triggered via Docker (no matter if on GitHub CI or locally), the `DAEMON_VERSION` can be set automatically. Local command would be
```
$ docker build -o build-output --build-arg "GIT_DESCRIPTION=$(git describe --tags)" .
```
If you prefer some other hard-coded version, you can use `--build-arg DAEMON_VERSION=...` instead.
So nothing should "remain broken". If you build without docker, no replacement is done (so behavior depends on the value in `constants.py`).


> The releases will have AppImages with correct versions, and sources with the correct versions too as per their release tags?

I'm a little bit confused by the terms "releases", "versions" and "sources". The version of a GitHub Release is taken from a git tag. The version of an AppImage is [also derived from git](https://github.com/C0rn3j/sc-controller/blob/037a5d65d11aa6358ea5cc4c788bb0c64682fbf4/.github/workflows/appimage.yml#L79), it just adds a classifier for the distro. The source code is tagged in Git. So yes, everything matches and Git is the only location where the version is managed.


> EDIT: Ah it does not touch the release files -> could it? Then we could have it fixed everywhere without bumping constants. We could have some soft-error check in the CI that will warn about the version being out of date in the repo for those.

Sorry, I don't understand. There should be no chance for anything to get out-of-date.


> Or perhaps use tags for source builds if it matches the latest, and git hash for source builds where there's no direct tag for the commit.

The `git describe` command applies this logic. When you build and the git `HEAD` points to a tag, the tag is used as version. When `HEAD` does not point to a tag, a new version is calculated from the last tag and the number of commits since that one.


Do these explanations help a little bit? Do you have further questions or see problems with this approach? Does the code contain any other version numbers besides the `DAEMON_VERSION`?

**Comment by C0rn3j** [comment]
>  Does the code contain any other version numbers besides the DAEMON_VERSION?

I have *just* [mostly migrated](https://github.com/C0rn3j/sc-controller/commit/b3edd804552e72d8960439a7942e183fc5a24768) from the deprecated `setup.py` to `pyproject.toml` (and threw `ruff.toml` into it as a bonus).

This let me fix the forever-somewhat-broken About menu, but it does mean that we have a new addition in the form of `[project]->version`.

Unsure how to deal with it and at the moment am exhausted, will give your comment proper attention later.

**Comment by git-developer** [comment]
It seems a lot of things are involved now. I'll focus on the version here.

I think that we should have a single location that contains the version. Currently, this is `constants.py`. Maybe it's possible to have a dedicated file that contains nothing but the version, and that is used to fill `DAEMON_VERSION`?

Regarding the build setup changes: It [seems to be possible](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html) to use the version from git by setting `dynamic = ["version"]` in the toml file without hard-coding it. Maybe that is an option?

**Comment by git-developer** [comment]
Off-topic: current build failure might be related to the python version: it fails on bullseye, focal, jammy (python <3.11) and works on bookworm, trixie, noble (python >= 3.11). If there's more to discuss, you could open a dedicated issue or enable _Discussions_ for this GitHub project.

**Comment by C0rn3j** [comment]
Thaaat was the wrong number.

**Comment by C0rn3j** [comment]
So, with recent changes, we have version in `constants.py` AND `pyproject.toml`.  
It definitely does not make sense to have both, and it makes more sense to keep it in `pyproject.toml`.

Now, we have these use cases:

* run.sh -> Build env in its own venv - until last version it ran itself from the source files
* Docker -> Build env
* AppImages
  * Ran from Git CI -> Build env
  * Ran from `scripts/appimage-AppRun.sh` -> Does this even work at all? It looks to hardcode a bunch of old versions.
* Distro packages
  * Built from a stable tag/release commit
  * Built from latest git master or otherwise not a direct tag/release
 
I see you've covered pretty much everything already.

I would change the versioning though, current one it not clear as to what commit it is:

`v0.4.9.1.1-8-g9f7dc8b`
\>\>\>
`0.4.9.1.1.8.dev167238795`

I would suggest one of these formats instead if it's an out-of-tag version, so we get to keep the exact commit hash:  
`v0.4.9.1.1-8-g9f7dc8b`
\>\>\>
`v0.4.9.1.1.dev8.g9f7dc8b`
`v0.4.9.1.1.8.g9f7dc8b`

**Comment by git-developer** [comment]
Use cases are simpler IMHO. AppImage just calls Docker and doesn't need anything (uses Git Metadata already). `scripts/appimage-AppRun.sh` and `appimage-build.sh` are both obsolete, coming from the pre-AppImageBuilder era. When TOML can be configured for `dynamic`, there's only `constants.py` remaining that needs the special handling from this PR.

> it not clear as to what commit it is

Have a look at the code: the `dev` number is hex-to-dec converted from the git hash. Looks ugly in my eyes, but is a 1:1 representation of the git hash. Your supposed version schemes unfortunately are not compatible to PEP 440 and will not be accepted by tools as version number for a python package. See the link from my first post for the spec. In short, only dotted numbers are valid. That's why I hacked the hash-to-number conversion into it. But after some thinking about it, I now think the conversion has no benefit: it should be OK to drop the hash and just suffix `.dev1` or so.

**Comment by C0rn3j** [comment]
> In short, only dotted numbers are valid

I skimmed it and my understanding was that a-Z 0-9 + dots was valid

**Comment by git-developer** [comment]
> Public version identifiers are separated into up to five segments:
>  - Epoch segment: N!
>   - Release segment: N(.N)*
>   - Pre-release segment: {a|b|rc}N
>  - Post-release segment: .postN
>  - Development release segment: .devN

Only local versions allow letters, separated by a +

**Comment by git-developer** [comment]
Python module _packaging_ has a Version class that validates the allowed patterns, you can play with it to check.

**Comment by git-developer** [comment]
BTW, I just noticed that in the upstream repo are already a bunch of 0.4.9.* [pre-releases](https://github.com/kozec/sc-controller/releases), the latest is [0.4.9.12](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.12). Maybe we should change to 0.4.10, 0.5 or 1.0.

**Comment by git-developer** [comment]
Let me explain. Suppose that the last release is `v0.4.8.21` and we're developing in branch `python3` containing 2 commits and we have another branch `fix/foo` with 1 commit:
```
+--------------------------o
|                          ^
o-----o-----o              |
^           ^              |
v0.4.8.21   python3        fix/foo
```

`git describe --tags` produces:
* for branch `python3`: `v0.4.8.21-2-g1a2b3c` (**2** commit since the last tag + commit hash)
* for branch `fix/foo`:  `v0.4.8.21-1-g4d5e6f` (**1** commit since the last tag + commit hash)

Currently implemented daemon version:
* for branch `python3`: `0.4.8.21.2.dev1715004` (whereas hex `1a2b3c` is dec `1715004`)
* for branch `fix/foo`:  `0.4.8.21.1.dev5070447`

**Comment by git-developer** [comment]
Related: https://stackoverflow.com/questions/35522047/pep440-compliant-and-git-describe-info-available-from-deployed-package

**Comment by git-developer** [comment]
OK, after reading the SO answers I think that `v0.4.8.21-2-g1a2b3c` should be converted to `0.4.8.21.2+g1a2b3c`. This _local_ number should be allowed as long as such a release is not published to a packaging repo.

The `dev` suffix is not appropriate because
1. > Developmental releases are ordered by their numerical component,
2. `X.dev0` is before `X`, not after.

**Comment by C0rn3j** [comment]
> BTW, I just noticed that in the upstream repo are already a bunch of 0.4.9.* [pre-releases](https://github.com/kozec/sc-controller/releases), the latest is [0.4.9.12](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.12). Maybe we should change to 0.4.10, 0.5 or 1.0.

I am starting to think we should stop caring about upstream as the project appears pretty much dead(kozec said he does not have access to his account anymore and does not seem on caring to restore it or create a new one), and just do whatever versioning we want.

0.4.10 at minimum is a good idea to do after I am done breaking things constantly with the beta releases.

**Comment by git-developer** [comment]
The last commit removes all the manual version fiddling and delegates to `setuptools-scm`. It determines a version from git similiar to the ideas above and generates a file with a python variable that can be used within the code. I tried here locally, works fine. Do you agree that this is a better way to go?

This approach has one downside: setuptools must be run for the version to exist. This breaks the tests currently because I don't know how to tell them about the version. Could you have a look? Maybe the `import` should not be done in `constants.py`, or additionally somewhere in the test code?

**Comment by git-developer** [comment]
I figured it out. This PR now contains the following changes:

* Version number is calculated by `setuptools-scm`
* `DAEMON_VERSION` is set accordingly
* AppImage version is set accordingly
* Python tests have been added to the Docker build
* Python dependency `email` has been removed from excludes to avoid a crash when the About dialog is opened

What's left for here or seperate PR(s):
* Remove obsolete appimage scripts
* Check if we really need 2 GitHub builds

File: Dockerfile [hunk 1]
```diff
@@ -14,12 +14,16 @@ RUN <<EOR
...
+	export DEBIAN_FRONTEND=noninteractive
 	apt-get install -y --no-install-recommends \
 		gcc \
+		git \
 		librsvg2-bin \
+		libxfixes3 \
 		linux-headers-generic \
 		python3-dev \
 		python3-setuptools \
+		python3-usb \
```
Comment (C0rn3j) [inline Dockerfile:26]: libusb1 is installed later with pip, is this necessary?

Comment (git-developer) [reply Dockerfile:26]: If removed, Docker build fails on python tests with
```
OSError: libusb-1.0.so: cannot open shared object file: No such file or directory
```

**Comment by git-developer** [comment]
Focal AppImage does not show About dialog:
```
Traceback (most recent call last):
  File "/tmp/.mount_sc-conef5HTv/usr/lib/python3.8/site-packages/scc/gui/app.py", line 870, in on_mnuAbout_activate
    AboutDialog(self).show(self.window)
  File "/tmp/.mount_sc-conef5HTv/usr/lib/python3.8/site-packages/scc/gui/aboutdialog.py", line 12, in __init__
    self.setup_widgets()
  File "/tmp/.mount_sc-conef5HTv/usr/lib/python3.8/site-packages/scc/gui/aboutdialog.py", line 30, in setup_widgets
    sccontroller_module = importlib.resources.files("scc")
AttributeError: module 'importlib.resources' has no attribute 'files'
```

`importlib.resources.files` was [added in Python 3.9](https://docs.python.org/3/library/importlib.resources.html#importlib.resources.files). Can we workaround this?

All other AppImages work fine.

**Comment by C0rn3j** [comment]
Is there anything we get out of supporting focal, outside of having to keep multiple focal-specific workarounds?

As far as I understand, the testing has shown that everything that works on focal also works on jammy.  
Considering Python 3.8 is EOL at the end of this month, we could simply drop distributions that do not ship 3.9+ (which I believe is only focal as bullseye is 3.9) and bump minimum Python req to 3.9.

Other than that, we could do some awful ImportError trycatch.

**Comment by git-developer** [comment]
I agree that it's not worth to put efforts into supporting non-critical features for a distro that is EOL soon. So maybe it's best to leave the About dialog broken in Focal and prepare a PR for removal of Focal in 04/25.

**Comment by C0rn3j** [comment]
Let's deal with the rest of things elsewhere, thanks a lot for implementing this!

**Comment by C0rn3j** [comment]
I have ended up bumping the minimum Python version to 3.9 in the pyproject.toml and README.  
It should not change much of anything, as all of the annoying typing changes we keep working around are only implemented in 3.10.

`focal` images were kept with a warning in the release notes about the broken About dialogue.

**Pull request merged**